library/std/src/sys/unix/ext/net/ancillary.rs

```diff
@@ -0,0 +1,749 @@
+        let mut previous_cmsg = cmsg;
+        while !cmsg.is_null() {
+            previous_cmsg = cmsg;
+            cmsg = libc::CMSG_NXTHDR(&msg, cmsg);
+        }
+
+        if previous_cmsg.is_null() {
+            return false;
+        }
+
+        (*previous_cmsg).cmsg_level = cmsg_level;
+        (*previous_cmsg).cmsg_type = cmsg_type;
+        (*previous_cmsg).cmsg_len = libc::CMSG_LEN(len) as usize;
+
+        let data = libc::CMSG_DATA(previous_cmsg).cast();
+
+        libc::memcpy(data, source.as_ptr().cast(), len as usize);
+    }
+    true
+}
+
+struct AncillaryDataIter<'a, T> {
+    data: &'a [u8],
+    phantom: crate::marker::PhantomData<T>,
+}
+
+impl<'a, T> AncillaryDataIter<'a, T> {
+    pub fn new(data: &'a [u8]) -> AncillaryDataIter<'a, T> {
+        AncillaryDataIter { data, phantom: PhantomData }
+    }
+}
+
+impl<'a, T> Iterator for AncillaryDataIter<'a, T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<T> {
+        unsafe {
+            let mut unit = zeroed();
+            if size_of::<T>() <= self.data.len() {
+                let unit_ptr: *mut T = &mut unit;
+                libc::memcpy(unit_ptr.cast(), self.data.as_ptr().cast(), size_of::<T>());
+                self.data = &self.data[size_of::<T>()..];
+                Some(unit)
+            } else {
+                None
+            }
+        }
+    }
+}
+
+#[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+#[derive(Clone)]
+pub struct UCred(libc::ucred);
+
+impl UCred {
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn new() -> UCred {
+        UCred(libc::ucred { pid: 0, uid: 0, gid: 0 })
+    }
+
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn set_pid(&mut self, pid: i32) {
+        self.0.pid = pid;
+    }
+
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn get_pid(&self) -> i32 {
+        self.0.pid
+    }
+
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn set_uid(&mut self, uid: u32) {
+        self.0.uid = uid;
+    }
+
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn get_uid(&self) -> u32 {
+        self.0.uid
+    }
+
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn set_gid(&mut self, gid: u32) {
+        self.0.gid = gid;
+    }
+
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn get_gid(&self) -> u32 {
+        self.0.gid
+    }
+}
+
+#[cfg(any(
+    target_os = "haiku",
+    target_os = "solaris",
+    target_os = "illumos",
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "freebsd",
+    target_os = "dragonfly",
+    target_os = "openbsd",
+    target_os = "netbsd",
+    target_os = "linux",
+    target_os = "android",
+    target_os = "emscripten",
+    target_os = "fuchsia",
+    target_env = "uclibc",
+))]
```

Comment:
This is the first instance I believe where we expose from `std::os::unix` public types that are conditional on more than just `cfg(unix)`. This is a major shift in how platform-specific standard library APIs are exposed.

If we decide to keep these the way they are, these need `doc(cfg(...))` annotations too.

Reply:
I do not understand. I never used `doc(cfg(...))`. Can you show me an example?

Reply:
It would be nice to sort the list alphabetically, so it's easier to check for a target at a glance.

Reply:
I sorted the list alphabetically (de2fa99e3e038b3ff2eda82891678476b939bfdc).

Reply:
@LinkTed The idea is that when generating documentation, you want to generate *all* the documentation, and tag it by where you can use it.

To do that, you need to do two things:

- Make sure that every `cfg` that would restrict something to specific platforms also allows it if `doc`; for instance, `cfg(any(doc, target_os = "...", ...))`. That way, it'll always be visible to rustdoc.

- Add `doc(cfg(...))` to such items (or to whole modules where appropriate), which tells rustdoc the configuration options needed for a given item.

Together, this is what generates the target annotations on pages like https://doc.rust-lang.org/std/os/unix/fs/trait.OpenOptionsExt.html and https://doc.rust-lang.org/std/os/windows/fs/trait.OpenOptionsExt.html .

Reply:
Thanks for the explanation.